src/System.Text.Primitives/System/Text/Encoding/Utf16/Utf16LittleEndianEncoder.cs

```diff
@@ -2,102 +2,204 @@
+        private const uint Utf16LowSurrogateLastCodePoint = 0xDFFF;
+
+        private const uint Utf16SurrogateRangeStart = Utf16HighSurrogateFirstCodePoint;
+        private const uint Utf16SurrogateRangeEnd = Utf16LowSurrogateLastCodePoint;
+
+        // To get this to compile with dotnet cli, we need to temporarily un-binary the magic values
+        private const byte b0000_0111U = 7;
+        private const byte b0000_1111U = 15;
+        private const byte b0001_1111U = 31;
+        private const byte b0011_1111U = 63;
+        private const byte b0111_1111U = 127;
+        private const byte b1000_0000U = 128;
+        private const byte b1100_0000U = 192;
+        private const byte b1110_0000U = 224;
+        private const byte b1111_0000U = 240;
+        private const byte b1111_1000U = 248;
+
+        #endregion Encoding Constants
+
+        #region Decoding implementation
+
+        /// <summary>
+        /// Decodes a span of UTF-16 characters into UTF-32.
+        /// 
+        /// This method will consume as many of the input characters as possible.
+        ///
+        /// On successful exit, the entire input was consumed and encoded successfully. In this case, <paramref name="charactersConsumed"/> will be
+        /// equal to the length of the <paramref name="utf16"/> and <paramref name="bytesWritten"/> will equal the total number of bytes written to
+        /// the <paramref name="utf32"/>.
+        /// 
+        /// On unsuccessful exit, the following conditions can exist.
+        ///  1) If the output buffer has been filled and no more input characters can be encoded, another call to this method with the input sliced to
+        ///     exclude the already encoded characters (using <paramref name="charactersConsumed"/>) and a new output buffer will continue the encoding.
+        ///  2) Encoding may have also stopped because the input buffer contains an invalid sequence.
+        /// </summary>
+        /// <param name="utf16">A span containing a sequence of UTF-16 characters.</param>
+        /// <param name="utf32">A span to write the UTF-32 data into.</param>
+        /// <param name="charactersConsumed">On exit, contains the number of code points that were consumed from the UTF-16 character span.</param>
+        /// <param name="bytesWritten">An output parameter to store the number of bytes written to <paramref name="utf32"/></param>
+        /// <returns>True if the input buffer was fully encoded into the output buffer, otherwise false.</returns>
+        public static bool TryDecode(ReadOnlySpan<char> utf16, Span<uint> utf32, out int charactersConsumed, out int charactersWritten)
         {
-            if (buffer.Length < 2)
-            {
-                codePoint = default(UnicodeCodePoint);
-                encodedBytes = default(int);
-                // buffer too small
+            charactersConsumed = 0;
+            charactersWritten = 0;
+
+            // TODO: Is this really what we want?
```

Comment:
I originally did that, but it broke some test cases. I made the assumption that it was desired.

I agree with you. I'll change it to succeed when input is empty and fix the tests.